src/circuit_writer/ir.rs

```diff
@@ -987,28 +987,39 @@ impl<B: Backend> IRWriter<B> {
                     .ok_or_else(|| self.error(ErrorKind::ExpectedConstant, expr.span))?;
                 let idx: usize = idx.try_into().unwrap();
 
-                // retrieve the type of the elements in the array
-                let array_typ = self.expr_type(array).expect("cannot find type of array");
+                // retrieve the type of the elements in the container
+                let container_typ = self
+                    .expr_type(container)
+                    .expect("cannot find type of container");
 
-                let elem_type = match array_typ {
+                let elem_type = match container_typ {
                     TyKind::Array(ty, array_len) => {
                         if idx >= (*array_len as usize) {
                             return Err(self.error(
                                 ErrorKind::ArrayIndexOutOfBounds(idx, *array_len as usize - 1),
                                 expr.span,
                             ));
                         }
-                        ty
+                        (**ty).clone()
+                    }
+                    TyKind::Tuple(typs) => {
+                        if idx >= typs.len() {
+                            return Err(self.error(
+                                ErrorKind::TupleIndexOutofBounds(idx, typs.len()),
+                                expr.span,
+                            ));
+                        }
+                        typs[idx].clone()
                     }
                     _ => Err(Error::new(
                         "compute-expr",
-                        ErrorKind::UnexpectedError("expected array"),
+                        ErrorKind::UnexpectedError("expected container"),
                         expr.span,
                     ))?,
                 };
 
-                // compute the size of each element in the array
-                let len = self.size_of(elem_type);
+                // compute the size of each element in the container
+                let len = self.size_of(&elem_type);
```

Comment:
Same issue I mentioned in the circuit writer.
The start index for the tuple doesn't look correct.